src/include/migraphx/matcher.hpp

```diff
@@ -857,23 +857,56 @@ auto skip_broadcasts_converts(Ms... ms)
     return skip(name("broadcast", "multibroadcast", "contiguous", "convert"))(ms...);
 }
 
-template <class T>
-inline auto has_value(T x, float tolerance = 1e-6)
+template <class F>
+inline auto literal_value_checker(F f)
 {
     return skip_broadcasts_converts(make_basic_pred_matcher([=](instruction_ref ins) {
         if(ins->name() != "@literal")
             return false;
         auto l = ins->get_literal();
         if(l.empty())
             return false;
+        return f(l);
+    }));
+}
+
+/**
+ * Uses integer multiples of the corresponding floating point epsilon and
+ * compares with abs(y - x) < eps * (atol_mult + rtol_mult * abs(y)).
+ * atol_mult controls the absolute tolerance.
+ * rtol_mult controls the relative tolerance.
+ * Uses no tolerance for integral types.
+ */
+template <class T>
+inline auto has_value(T x, std::size_t atol_mult = 10, std::size_t rtol_mult = 10)
+{
+    return literal_value_checker([=](migraphx::literal l) {
         bool b = false;
         l.visit([&](auto v) {
-            if(std::all_of(
-                   v.begin(), v.end(), [&](auto val) { return std::fabs(val - x) < tolerance; }))
-                b = true;
+            // cast to the literal's data type before comparing
+            using type = typename decltype(v)::value_type;
+            if constexpr(std::is_integral<type>{})
+            {
+                if(std::all_of(
+                       v.begin(), v.end(), [&](auto val) { return val == static_cast<type>(x); }))
+                {
+                    b = true;
+                }
+            }
+            else
+            {
+                auto eps = std::numeric_limits<type>::epsilon();
+                if(std::all_of(v.begin(), v.end(), [&](auto val) {
+                       return std::fabs(val - static_cast<type>(x)) <
```

Comment:
Comment: All values inside literal are compare against `x`.  That would probably mean all values inside literal `y` are of same value.

Reply:
We dont know if they are all the same value, that is why we check that they are the same value.